java.lang.IllegalStateException: Too many receivers, total of 1000

<!-- DO NOT DELETE 
validate_template=true
template_path=.github/ISSUE_TEMPLATE/bug.md
-->
### Environment

  * Android Studio version: 3.6.3
  * Firebase Component: Firebase Crashlytics, Firebase Analytics, Firebase Remote Config
  * Component version: 17.2.2, 18.0.0, 20.0.0

Additionally:
```
        classpath 'com.android.tools.build:gradle:3.6.4'
        classpath 'com.google.gms:google-services:4.3.4'
        classpath 'com.google.android.gms:strict-version-matcher-plugin:1.2.2'
        classpath 'com.google.firebase:firebase-crashlytics-gradle:2.3.0'
```

```
distributionUrl=https\://services.gradle.org/distributions/gradle-5.6.4-all.zip
```

### The problem

After trying to upgrade from Fabric Crashlytics to Firebase Crashlytics (published only that change), our users experienced very severe issue and here is the log:

```
java.lang.RuntimeException: 
  at android.app.ActivityThread.handleCreateService (http://ActivityThread.java:4096)
  at android.app.ActivityThread.access$1900 (http://ActivityThread.java:232)
  at android.app.ActivityThread$H.handleMessage (http://ActivityThread.java:1982)
  at android.os.Handler.dispatchMessage (http://Handler.java:107)
  at android.os.Looper.loop (http://Looper.java:214)
  at android.app.ActivityThread.main (http://ActivityThread.java:7710)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (http://RuntimeInit.java:516)
  at com.android.internal.os.ZygoteInit.main (http://ZygoteInit.java:950)
Caused by: java.lang.IllegalStateException: 
  at android.os.Parcel.createException (http://Parcel.java:2079)
  at android.os.Parcel.readException (http://Parcel.java:2039)
  at android.os.Parcel.readException (http://Parcel.java:1987)
  at android.app.IActivityManager$Stub$Proxy.registerReceiver (http://IActivityManager.java:4860)
  at android.app.ContextImpl.registerReceiverInternal (http://ContextImpl.java:1561)
  at android.app.ContextImpl.registerReceiver (http://ContextImpl.java:1512)
  at android.app.ContextImpl.registerReceiver (http://ContextImpl.java:1500)
  at android.content.ContextWrapper.registerReceiver (http://ContextWrapper.java:627)
  at ule.android.cbc.ca.listenandroid.notification.MediaNotification.registerNotificationReceiver (http://MediaNotification.java:191)
  at ule.android.cbc.ca.listenandroid.services.MediaService.onCreate (http://MediaService.java:318)
  at android.app.ActivityThread.handleCreateService (http://ActivityThread.java:4084)
Caused by: android.os.RemoteException: 
  at com.android.server.am.ActivityManagerService.registerReceiver (http://ActivityManagerService.java:15759)
  at android.app.IActivityManager$Stub.onTransact (http://IActivityManager.java:1923)
  at com.android.server.am.ActivityManagerService.onTransact (http://ActivityManagerService.java:3021)
  at android.os.Binder.execTransactInternal (http://Binder.java:1032)
  at android.os.Binder.execTransact (http://Binder.java:1005)
```

It only happens to a certain percentage of our users, but the ones who have it it's super severe (on 10 users we have about 70 crashes). That low percentage ends up being very big, since our overall audience number is around 250K.

Also a thing to note, it only happens on specific devices like: LG, Xiaomi, OnePlus,.. Never on Google Pixel.

#### Steps to reproduce

We actually only know this happens in the background. We're currently in the process of acquiring more devices so we can try to reproduce it, but on our devices this has never happened.

--- a/firebase-crashlytics/src/androidTest/java/com/google/firebase/crashlytics/internal/common/BatteryStateTest.java
+++ b/firebase-crashlytics/src/androidTest/java/com/google/firebase/crashlytics/internal/common/BatteryStateTest.java
@@ -58,6 +58,18 @@ public class BatteryStateTest extends CrashlyticsTestCase {
     assertEquals(1, state.getBatteryVelocity());
   }
 
+  public void testTooManyReceivers() {
+    Context mockContext = mock(Context.class);
+    when(mockContext.registerReceiver(isNull(), any()))
+        .thenThrow(new IllegalStateException("Too many receivers"));
+
+    BatteryState state = BatteryState.get(mockContext);
+
+    assertNull(state.getBatteryLevel());
+    assertFalse(state.isPowerConnected());
+    assertEquals(1, state.getBatteryVelocity());
+  }
+
   public void testEmptyIntent() {
     final Context mockContext = mock(Context.class);
     when(mockContext.registerReceiver(isNull(), any())).thenReturn(new Intent());

--- a/firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/common/BatteryState.java
+++ b/firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/common/BatteryState.java
@@ -18,6 +18,7 @@ import android.content.Context;
 import android.content.Intent;
 import android.content.IntentFilter;
 import android.os.BatteryManager;
+import com.google.firebase.crashlytics.internal.Logger;
 
 /** A utility class representing the state of the battery. */
 class BatteryState {
@@ -66,11 +67,16 @@ class BatteryState {
     boolean powerConnected = false;
     Float level = null;
 
-    final IntentFilter ifilter = new IntentFilter(Intent.ACTION_BATTERY_CHANGED);
-    final Intent batteryStatusIntent = context.registerReceiver(null, ifilter);
-    if (batteryStatusIntent != null) {
-      powerConnected = isPowerConnected(batteryStatusIntent);
-      level = getLevel(batteryStatusIntent);
+    try {
+      final IntentFilter ifilter = new IntentFilter(Intent.ACTION_BATTERY_CHANGED);
+      final Intent batteryStatusIntent = context.registerReceiver(/*receiver=*/ null, ifilter);
+      if (batteryStatusIntent != null) {
+        powerConnected = isPowerConnected(batteryStatusIntent);
+        level = getLevel(batteryStatusIntent);
+      }
+    } catch (IllegalStateException ex) {
+      // This happens on some devices when the app registers too many receivers.
+      Logger.getLogger().e("An error occurred getting battery state.", ex);
     }
 
     return new BatteryState(level, powerConnected);

